File: Example/OpAdxSdkDemo_OC/OpAdxDemo_OC/Classes/RewardedInterstitialViewController.m
//
//  RewardedInterstitialViewController.m
//  OpAdxSdkDemo_OC
//
//  Created by Luan Chen on 2026/1/26.
//

#import "RewardedInterstitialViewController.h"
#import <OpAdxSdk/OpAdxSdk.h>

@interface RewardedInterstitialViewController ()<OpAdxRewardedInterstitialAdDelegate>

@property (nonatomic, strong, nullable) OpAdxRewardedInterstitialAdBridge *rewardedInterstitialAd;

@end

@implementation RewardedInterstitialViewController

- (BOOL)hasVideo {
    return YES;
}

- (AdFormat)adFormat {
    return AdFormatRewardedInterstitial;
}

- (NSString *)adFormatString {
    return @"RewardedInterstitial Ad";
}

- (void)loadAd {
    if (!self.placementId) return;
    
    [self.logView print:@"Loading ..."];
    if (self.rewardedInterstitialAd != nil) {
        [self destroyAd];
    }
    
    OpAdxRewardedInterstitialAdBridge *rewardedInterstitialAd = [[OpAdxRewardedInterstitialAdBridge alloc] initWithPlacementId:self.placementId auctionType: AdAuctionTypeRegular];
    self.rewardedInterstitialAd = rewardedInterstitialAd;
    rewardedInterstitialAd.delegate = self;
    [rewardedInterstitialAd loadAd];
}

- (void)showAd {
    if (!self.rewardedInterstitialAd) return;
    
    if (![self.rewardedInterstitialAd isAdValid]) {
        [self.logView print:@"Ad is invalidated."];
        [self destroyAd];
        return;
    }
    
    __weak typeof(self) weakSelf = self;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [weakSelf.rewardedInterstitialAd showAdFrom:weakSelf];
    });
    
    [self enableDestroyAd];
    [self disableShowAd];
}

- (void)destroyAd {
    [super destroyAd];
    self.rewardedInterstitialAd = nil;
    [self disableDestroyAd];
    [self disableShowAd];
}

#pragma mark - OpAdxRewardedInterstitialAdDelegate

- (void)rewardedInterstitialAdDidLoad:(OpAdxRewardedInterstitialAdBridge *)rewardedInterstitialAd {
    NSLog(@"[ADX] RewardedInterstitial广告加载成功: %@", rewardedInterstitialAd.placementID);
    [self.logView print:[NSString stringWithFormat:@"Loaded."]];
   
    [self enableShowAd];
    [self enableDestroyAd];
    
}

- (void)rewardedInterstitialAd:(OpAdxRewardedInterstitialAdBridge *)rewardedInterstitialAd didFailWithError:(OpAdxAdError *)error {
    NSLog(@"[ADX] RewardedInterstitial广告加载失败: %@ %@", rewardedInterstitialAd.placementID, error);
    [self.logView print:error.message];
    
}

- (void)rewardedInterstitialAdDidClick:(OpAdxRewardedInterstitialAdBridge *)rewardedInterstitialAd {
    NSLog(@"[ADX] RewardedInterstitial广告被点击: %@", rewardedInterstitialAd.placementID);
    [self.logView print:@"onAdDidClick"];
    
}

- (void)rewardedInterstitialAdDidClose:(OpAdxRewardedInterstitialAdBridge *)rewardedInterstitialAd {
    NSLog(@"[ADX] RewardedInterstitial广告关闭: %@", rewardedInterstitialAd.placementID);
    [self.logView print:@"onAdDidClose"];
    // 清理
    self.rewardedInterstitialAd = nil;

}

- (void)rewardedInterstitialAdWillLogImpression:(OpAdxRewardedInterstitialAdBridge *)rewardedInterstitialAd {
    NSLog(@"[ADX] RewardedInterstitial广告展示: %@", rewardedInterstitialAd.placementID);
    [self.logView print:@"onAdImpression"];
    
}

@end

File: Example/OpAdxSdkDemo_Swift/OpAdxSdkDemo/RewardedInterstitialViewController.swift
//
//  RewardedInterstitialViewController.swift
//  OpAdxSdkDemo
//
//  Created by Luan Chen on 2026/1/19.
//

import UIKit
import OpAdxSdk

class RewardedInterstitialViewController: BaseViewController {
    
    private var rewardedInterstitialAd: OpAdxRewardedInterstitialAd?
    
    override var hasVideo: Bool {
        return false
    }
    
    override var adFormat: AdFormat {
        return .rewardedInterstitial
    }
    
    override var adFormatString: String {
        return "Rewarded Interstitial Ad"
    }
    
    override func loadAd() {
        guard let placementId = placementId else { return }
        
        logView.print("Loading ...")
        if rewardedInterstitialAd != nil {
            self.destroyAd()
        }
        
        rewardedInterstitialAd = OpAdxRewardedInterstitialAd(placementId: placementId, auctionType: AdAuctionType.regular
        )
        // 创建监听器实例
        let listener = OpAdxRewardedInterstitialAdLoadListenerImp(
            onAdLoaded: { [weak self] ad in
                guard let self = self else { return }
                self.logView.print("Loaded")
                self.enableShowAd()
                self.enableDestroyAd()
            },
            onAdFailedToLoad: { [weak self] error in
                guard let self = self else { return }
                self.logView.print(error.message)
            }
        )
        
        rewardedInterstitialAd?.load(placementId: placementId, listener: listener)
    }
    
    override func showAd() {
        guard let rewardedInterstitialAd = rewardedInterstitialAd else { return }
        
        if rewardedInterstitialAd.isAdInvalidated() {
            logView.print("Ad is invalidated.")
            destroyAd()
            return
        }
        
        // Set scene ID and SSV options
        // scene id: max length 100 bytes after url encoded, or will be discarded.
        rewardedInterstitialAd.setSceneId("Demo scene #2")
        
        // 创建 SSV 选项
        let rewardSsvOptions = RewardSsvOptions.Builder()
            // user id: max length 100 bytes after url encoded, or will be discarded.
            .userId("Demo user id %:{测试?}")
            // custom data: max length 1KB after url encoded, or will be discarded.
            .customData("Demo user custom data %:{测试?}#2")
            .build()
        
        rewardedInterstitialAd.setRewardSsvOptions(rewardSsvOptions)
        
        // 创建交互监听器
        let interactionListener = OpAdxRewardedInterstitialAdInteractionListenerImp(
            onAdClicked: { [weak self] in
                guard let self = self else { return }
                self.logView.print("Clicked!")
            },
            onAdDisplayed: { [weak self] in
                guard let self = self else { return }
                self.logView.print("Displayed!")
            },
            onAdDismissed: { [weak self] in
                guard let self = self else { return }
                self.logView.print("Dismissed")
                self.destroyAd()
            },
            onAdFailedToShow: { [weak self] error in
                guard let self = self else { return }
                self.logView.print(error.message)
            },
            onUserRewarded: { [weak self] reward in
                guard let self = self else { return }
                self.logView.print("rewarded: type=\(reward.type), amount=\(reward.amount)")
            }
        )
        
        rewardedInterstitialAd.show(on: self, listener: interactionListener)
        
        enableDestroyAd()
        disableShowAd()
    }
    
    override func destroyAd() {
        super.destroyAd()
        rewardedInterstitialAd?.destroy()
        rewardedInterstitialAd = nil
        logView.print("Destroyed...")
        disableDestroyAd()
        disableShowAd()
    }
    
    deinit {
        destroyAd()
    }
}
